pwndbg/dbg/gdb.py

```diff
@@ -30,22 +31,67 @@ def parse_and_eval(expression: str, global_context: bool) -> gdb.Value:
         return gdb.parse_and_eval(expression)
 
 
+T = TypeVar("T")
+
+
+class Selection(ContextManager[None], Generic[T]):
+    """
+    GDB has a lot of global state. Many of our queries require that we select a
+    given object globally before we make them. When doing that, we must always
+    be careful to return selection to its previous state before exiting. This
+    class automatically manages the selection of a single object type.
+    """
+
+    select: Callable[[T], None]
+    target: T
+    get_current: Callable[[], T]
+
+    current: T
+    restore: bool
+
+    def __init__(self, target: T, get_current: Callable[[], T], select: Callable[[T], None]):
+        """
+        Upon entrace to the `with` block, the element given by `target` will be
+        compared to the object returned by calling `get_current`. If they
+        compare different, the value previously returned by `get_current` is
+        saved, and the element given by `target` will be selected by passing it
+        as an argument to `select`, and, after execution leaves the `with`
+        block, the previously saved element will be selected in the same fashion
+        as the first element.
+
+        If the elements don't compare different, this is a no-op.
+        """
+
+        self.get_current = get_current
+        self.select = select
+        self.target = target
+        self.restore = False
+        self.current = None
+
+    @override
+    def __enter__(self) -> None:
```

Comment:
Wouldn't it be simpler to just use a function context manager for this? https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager

Reply:
Yeah, probably. It just didn't occur to me at the time of writing. Do you want me to change to it?